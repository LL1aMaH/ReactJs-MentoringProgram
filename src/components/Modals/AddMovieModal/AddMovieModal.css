.modal {
  color: var(--primary-font-color);
  background-color: var(--black-background-color);
  display: flex;
  flex-direction: column;
  padding: 3rem 4rem;
  width: 40rem;

  & svg {
    width: 3rem;
    height: auto;
    align-self: flex-end;
    cursor: pointer;
  }

  & p {
    font-size: 35px;
    font-weight: 100;
    cursor: default;
  }

  & .label {
    font-size: 15px;
    color: var(--primary-link-color);
    cursor: default;
  }

  & input {
    background-color: var(--primary-contrast-color);
    border: none;
    border-radius: 4px;
    width: 100%;
    height: 3rem;
    font-size: 15px;
    padding: 0 10px;
    margin-top: 0.5rem;
    &::placeholder {
      color: var(--text-secondary-color);
    }

    &:focus {
      outline: none;
      border: 1px solid var(--primary-link-color);
    }
  }

  & .buttons {
    padding-top: 3rem;
    display: flex;
    justify-content: flex-end;
  }

  & .button {
    height: 3rem;
    border: 1px solid var(--primary-link-color);
    border-radius: 4px;
    margin-left: 0.5rem;
    font-size: 15px;
    padding: 0 2rem;

    &:hover {
      background-color: var(--primary-link-color);
    }
  }

  & .field {
    margin-top: 2rem;
  }
}

.multiSelect {
  margin-top: 0.5rem;
}
