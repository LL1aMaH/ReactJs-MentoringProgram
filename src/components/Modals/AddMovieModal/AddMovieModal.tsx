import React, { useCallback, useState } from 'react';
import { Button } from 'Components/Button';
import { InputField } from 'Components/InputField';
import { Modal } from 'Components/Modal';

import { SelectField } from 'Components/NavBar/components/SelectField';

import Close from 'Assets/icons/close.svg';

import classnames from 'classnames';
import styles from './AddMovieModal.css';

export type AddMovieModalProps = {
  className?: string;
};

export type AddMovieModalComponentProps = {
  onCancel: () => void;
};

const AddMovieModalComponent = ({ onCancel }: AddMovieModalComponentProps) => (
  <div className={styles.modal}>
    <Close onClick={onCancel} />
    <p>ADD MOVIE</p>
    <InputField
      placeholder="Moana"
      label="TITLE"
      labelClassName={styles.label}
      className={styles.field}
    />
    <InputField
      placeholder="Select Data"
      label="RELEASE DATA"
      labelClassName={styles.label}
      className={styles.field}
    />
    <InputField
      placeholder="Movie URL here"
      label="MOVIE URL"
      labelClassName={styles.label}
      className={styles.field}
    />
    <SelectField
      options={['Crime', 'Documentary', 'Horror', 'Comedy']}
      onChange={(val) => console.log(val)}
      label="GENRE"
      containerClassName={styles.containerSelectField}
      className={styles.selectField}
      buttonClassName={styles.selectButton}
      optionListClassName={styles.optionList}
      optionClassName={classnames(styles.selectButton, styles.option)}
      labelClassName={styles.label}
    />
    <InputField
      placeholder="Overview text goes here"
      label="OVERVIEW"
      labelClassName={styles.label}
      className={styles.field}
    />
    <InputField
      placeholder="Runtime Text Goes here"
      label="RUNTIME"
      labelClassName={styles.label}
      className={styles.field}
    />
    <div className={styles.buttons}>
      <Button className={styles.button}>RESET</Button>
      <Button className={styles.button} onClick={onCancel}>
        SUBMIT
      </Button>
    </div>
  </div>
);

export const AddMovieModal = ({ className }: AddMovieModalProps): JSX.Element => {
  const [isOpen, setOpen] = useState(false);

  const handleClose = useCallback(() => {
    setOpen(!isOpen);
  }, [isOpen]);

  return (
    <>
      <Button onClick={handleClose} className={className}>
        + ADD MOVIE
      </Button>
      <Modal open={isOpen} onClose={handleClose}>
        <AddMovieModalComponent onCancel={handleClose} />
      </Modal>
    </>
  );
};
