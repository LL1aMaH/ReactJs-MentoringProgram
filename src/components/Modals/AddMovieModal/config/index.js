import { object, string, date, number, mixed } from 'yup';

const validateDate = (val) => {
  return val < new Date();
};
export const validationSchema = object({
  title: string().min(2, 'Too Short!').max(50, 'Too Long!').required('Required'),
  release_date: date().test('date', 'not real date', validateDate),
  poster_path: string().min(2, 'Too Short!').required('Required'),
  runtime: number().min(2, 'Too Short!').max(250, 'Too Long!').required('Required'),
  overview: string().min(5, 'Too Short!').required('Required'),
  genres: mixed().required('Required'),
});
