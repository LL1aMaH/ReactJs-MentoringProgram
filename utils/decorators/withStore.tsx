import React from 'react';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import originalStore from './configureStore';

type TComponentFunc = () => JSX.Element;

const withStore = (componentFunc: TComponentFunc, initialState = {}): JSX.Element => {
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares);
  const store = mockStore(initialState);
  return <Provider store={store}>{componentFunc()}</Provider>;
};

export const withOriginalStore = (componentFunc: TComponentFunc): JSX.Element => {
  return <Provider store={originalStore}>{componentFunc()}</Provider>;
};

export default withStore;
