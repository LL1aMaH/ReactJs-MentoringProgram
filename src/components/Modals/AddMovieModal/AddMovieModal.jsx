import React, {  useCallback, useState } from 'react';

import { Formik, Form } from 'formik';

import { useDispatch, useSelector } from 'react-redux';

import { InputField } from '../../InputField/InputField';
import { Button } from '../../Button/Button';
import { Modal } from '../../Modal/Modal';

import Close from '../../../../public/close.svg';
import { GENRES } from '../../../consts/const';

import styles from '../../../../styles/AddMovieModal.module.scss';
import { validationSchema } from './config';

// import { addMovie } from '../../../store/actions';

// import { getStartButton } from '../../../store/selectors';

import { MultiSelectField } from '../components/MultiSelectField/MultiSelectField';


const AddMovieModalComponent = ({
  onCancel,
  setShowSuccessfulAddMovieModal,
}) => {
  const [state, setState] = useState([]);
  const dispatch = useDispatch();

  const { activeButtonStart, sortStart } = useSelector(getStartButton);

  const handleClick = (item) => {
    setState([...item]);
  };

  const handleSubmitForm = (values, { setSubmitting }) => {
    const correctValues = Object.assign(values, { genres: state });
    correctValues.runtime = +correctValues.runtime;
    addMovie(correctValues, activeButtonStart, sortStart, dispatch);
    setSubmitting(false);
    setShowSuccessfulAddMovieModal((visible) => !visible);
    onCancel();
  };

  const initialValues = {
    title: '',
    release_date: '',
    overview: '',
    runtime: 90,
    genres: [''],
    poster_path: '',
  };

  return (
    <div className={styles.modal}>
      <Close onClick={onCancel} />
      <p>ADD MOVIE</p>

      <Formik
        initialValues={initialValues}
        onSubmit={handleSubmitForm}
        validationSchema={validationSchema}
      >
        {({ errors }) => {
          return (
            <Form>
              <InputField
                placeholder="Title"
                label="TITLE"
                labelClassName={styles.label}
                className={styles.field}
                name="title"
                id="title"
                error={errors.title}
              />
              <InputField
                placeholder="Select Data"
                label="RELEASE DATA"
                labelClassName={styles.label}
                className={styles.field}
                type='date'
                name="release_date"
                id="release_date"
                error={errors.release_date}
              />
              <InputField
                placeholder="Movie URL here"
                label="MOVIE URL"
                labelClassName={styles.label}
                className={styles.field}
                name="poster_path"
                id="poster_path"
                error={errors.poster_path}
              />
              <MultiSelectField
                options={GENRES}
                value={state}
                onChange={handleClick}
                placeholder="Select Genre"
                className={styles.multiSelect}
                label="GENRE"
                labelClassName={styles.label}
              />
              <InputField
                placeholder="Overview text goes here"
                label="OVERVIEW"
                labelClassName={styles.label}
                className={styles.field}
                name="overview"
                id="overview"
                error={errors.overview}
              />
              <InputField
                placeholder="Runtime Text Goes here"
                label="RUNTIME"
                labelClassName={styles.label}
                className={styles.field}
                name="runtime"
                id="runtime"
                error={errors.runtime}
              />
              <div className={styles.buttons}>
                <Button className={styles.button} type="reset">
                  RESET
                </Button>
                <Button className={styles.button} type="submit">
                  SUBMIT
                </Button>
              </div>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};

export const AddMovieModal = ({
  className,
  setShowSuccessfulAddMovieModal,
}) => {
  const [isOpen, setOpen] = useState(false);

  const handleClose = useCallback(() => {
    setOpen(!isOpen);
  }, [isOpen]);
  return (
    <>
      <Button onClick={handleClose} className={className} testId="addMovie">
        + ADD MOVIE
      </Button>
      <Modal open={isOpen} onClose={handleClose}>
        <AddMovieModalComponent
          onCancel={handleClose}
          setShowSuccessfulAddMovieModal={setShowSuccessfulAddMovieModal}
        />
      </Modal>
    </>
  );
};
