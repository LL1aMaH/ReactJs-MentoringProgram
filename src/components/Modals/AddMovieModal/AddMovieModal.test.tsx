import React from 'react';

import { render, fireEvent } from '@testing-library/react';

import getFormDecorator from '../../../../utils/decorators/withForm';
import withStore from '../../../../utils/decorators/withStore';

import { AddMovieModal } from './AddMovieModal';

describe('AddMovieModal', () => {
  const withForm = getFormDecorator({ initialValues: { test: 'test' } });
  const setShowSuccessfulAddMovieModal = jest.fn;
  it('renders correctly', () => {
    const getComponent = () =>
      render(
        withForm(() => (
          <AddMovieModal setShowSuccessfulAddMovieModal={setShowSuccessfulAddMovieModal} />
        )),
      );

    expect(getComponent().asFragment()).toMatchSnapshot();
  });

  it('renders modal correctly', () => {
    const component = render(
      withStore(
        () =>
          withForm(() => (
            <AddMovieModal setShowSuccessfulAddMovieModal={setShowSuccessfulAddMovieModal} />
          )),
        {
          genre: 'ALL',
          sort: 'RELEASE DATE',
        },
      ),
    );

    const button = component.getByTestId('addMovie');
    fireEvent.click(button);

    expect(component.asFragment()).toMatchSnapshot();
  });
});
