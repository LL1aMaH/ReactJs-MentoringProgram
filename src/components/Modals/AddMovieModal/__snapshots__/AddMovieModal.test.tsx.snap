// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieModal renders correctly 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="addMovie"
    tabindex="0"
  >
    + ADD MOVIE
  </button>
</DocumentFragment>
`;

exports[`AddMovieModal renders modal correctly 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="addMovie"
    tabindex="0"
  >
    + ADD MOVIE
  </button>
</DocumentFragment>
`;
